# Introduction to Data Science 
# Inclass Exercises for Lecture 22: Neural Networks II
*COMP 5360 / MATH 4100, University of Utah, http://datasciencecourse.net/*


### Installing TensorFlow

Instructions for installing TensorFlow are available at [the tensorflow install page](https://www.tensorflow.org/versions/r1.0/install/).

It is recommended that you use the command: 
```
pip install tensorflow
```


In [1]:
import tensorflow as tf
print(tf.__version__)


2.16.1


**Exercise 1:** Use TensorFlow to compute the derivative of $f(x) = e^x$ at $x=2$.

In [6]:
# Define the function f(x) = e^x
def f(x):
    return tf.exp(x)

# Define the value of x at which to compute the derivative
x = tf.constant(2.0)

# Use tf.GradientTape() to compute the derivative of f(x) at x=2
with tf.GradientTape() as tape:
    tape.watch(x)
    y = f(x)

# Compute the derivative of f(x) at x=2
dy_dx = tape.gradient(y, x)

print(dy_dx.numpy())


7.389056


**Exercise 2:** Use TensorFlow to find the minimum of the [Rosenbrock function](https://en.wikipedia.org/wiki/Rosenbrock_function): 
$$
f(x,y) = (x-1)^2 + 100*(y-x^2)^2.
$$


In [11]:
# Define the Rosenbrock function
def rosenbrock(x, y):
    return (1 - x)**2 + 100 * (y - x**2)**2

# Define x and y as TensorFlow variables
x = tf.Variable(0.0)
y = tf.Variable(0.0)

# Define the optimizer (Adam optimizer with a learning rate of 0.01)
optimizer = tf.keras.optimizers.Adam(learning_rate=0.01)

# Optimization loop
for i in range(1000):
    with tf.GradientTape() as tape:
        # Watch the variables
        tape.watch([x, y])
        # Calculate the value of the Rosenbrock function
        z = rosenbrock(x, y)
    # Calculate gradients
    gradients = tape.gradient(z, [x, y])
    # Apply gradients to update variables
    optimizer.apply_gradients(zip(gradients, [x, y]))
print(x.numpy())
print(y.numpy())


0.99999714
1.0000006


## Using a pre-trained network

There are many examples of pre-trained NN that can be accessed [here](https://www.tensorflow.org/api_docs/python/tf/keras/applications). 
These NN are very large, having been trained on giant computers using massive datasets. 

It can be very useful to initialize a NN using one of these. This is called [transfer learning](https://en.wikipedia.org/wiki/Transfer_learning). 


We'll use a NN that was pretrained for image recognition. This NN was trained on the  [ImageNet](http://www.image-net.org/) project, which contains > 14 million images belonging to > 20,000 classes (synsets). 

In [1]:
import tensorflow as tf
import numpy as np
from tensorflow.keras.preprocessing import image
from tensorflow.keras.applications import vgg16

**Exercise 3:** Use tf.keras.applications.VGG16 (the NN pre-trained on ImageNet) to classify at least two images not done in lecture. These can be images from the lecture folder or your own images. Report on the top five predicted classes and their corresponding probabilities. 

In [4]:
vgg_model = tf.keras.applications.VGG16(weights='imagenet',include_top=True)
#vgg_model.summary()


In [3]:
# image 1
img_path = 'images/brodie.jpeg'
img = image.load_img(img_path, target_size=(224, 224))

x = image.img_to_array(img)
x = np.expand_dims(x, axis=0)
x = vgg16.preprocess_input(x)

preds = vgg_model.predict(x)
print('Predicted:', vgg16.decode_predictions(preds, top=5)[0])

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 239ms/step
35363/35363 ━━━━━━━━━━━━━━━━━━━━ 0s 1us/step
Predicted: [('n01877812', 'wallaby', 0.19841017), ('n02106662', 'German_shepherd', 0.15015586), ('n02325366', 'wood_rabbit', 0.09528107), ('n02105412', 'kelpie', 0.08297235), ('n02097047', 'miniature_schnauzer', 0.050912224)]


The top prediction for brodie is wallably with 19.8% probability.

In [6]:
# image 2
img_path = 'images/layla1.jpeg'
img = image.load_img(img_path, target_size=(224, 224))

x = image.img_to_array(img)
x = np.expand_dims(x, axis=0)
x = vgg16.preprocess_input(x)

preds = vgg_model.predict(x)
print('Predicted:', vgg16.decode_predictions(preds, top=5)[0])

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 200ms/step
Predicted: [('n02123045', 'tabby', 0.44996402), ('n02124075', 'Egyptian_cat', 0.21365678), ('n02123159', 'tiger_cat', 0.09498075), ('n04265275', 'space_heater', 0.015217376), ('n02971356', 'carton', 0.013566178)]


The top prediction for layla is tabby with 45% probability.

**Exercise 4 (optional):** There are several [other pre-trained networks in Keras](https://github.com/keras-team/keras-applications). Try these!